common/immutable.js

```diff
@@ -0,0 +1,90 @@
+import Immutable from 'seamless-immutable';
+
+function defineMethod(object, methodName, method) {
+  Object.defineProperty(object, methodName, {
+    enumerable: false,
+    configurable: false,
+    writable: false,
+    value: method
+  });
+}
+
+function FakeImmutable(state) {
+  if (typeof state !== 'object') {
+    return state;
+  }
+
+  defineMethod(state, 'getIn', function(keys) {
+    let currentRoot = this;
+
+    for (let index in keys) {
+      const key = keys[index];
+
+      if (currentRoot[key] === undefined) {
+        return undefined;
+      }
+
+      currentRoot = currentRoot[key];
+    }
+
+    return currentRoot;
+  });
+
+  defineMethod(state, 'set', function(key, value) {
+    this[key] = value;
+    return this;
+  });
+
+  defineMethod(state, 'setIn', function(keys, value) {
+    const lastIndex = keys.length - 1;
+
+    keys.reduce((currentRoot, key, index) => {
+      if (index === lastIndex) {
+        currentRoot[key] = value;
+      } else if (currentRoot[key] === undefined) {
+        currentRoot[key] = {};
+      }
+
+      return currentRoot[key];
+    }, this);
+
+    return this;
+  });
+
+  defineMethod(state, 'update', function(key, fn) {
+    const value = FakeImmutable(this.getIn([key]));
+    return this.set(key, fn(value));
+  });
+
+  defineMethod(state, 'updateIn', function(keys, fn) {
+    const value = FakeImmutable(this.getIn(keys));
+    return this.setIn(keys, fn(value));
+  });
+
+  defineMethod(state, 'without', function(key) {
+    if (!Array.isArray(key)) {
+      delete this[key];
```

Comment:
Refactor deleting array key to splice function. Instead, key will remained with undefined value